server/src/main/java/io/crate/expression/tablefunctions/GenerateSubscripts.java

```diff
@@ -0,0 +1,188 @@
+    }
+
+    private static List<?> checkAllElementsAreArraysOfConsistentDimension(List<?> array, int level) {
+        int dim = -1;
+        List<?> firstNonNullElement = null;
+        for (int i = 0; i < array.size(); i++) {
+            Object oi = array.get(i);
+            if (oi != null && !(oi instanceof List)) {
+                throw new IllegalArgumentException(String.format(
+                    Locale.ENGLISH,
+                    "expected a nested array, or null value at level %d, position %d",
+                    level, i + 1));
+            }
+            if (oi == null) {
+                // null is a valid value within an array
+                continue;
+            }
+            List<?> element = (List<?>) oi;
+            if (dim == -1) {
+                dim = element.size();
+            } else {
+                if (dim != element.size()) {
+                    throw new IllegalArgumentException(String.format(
+                        Locale.ENGLISH,
+                        "nested arrays must have the same dimension within a level, offending level %d, position %d",
+                        level, i + 1));
+                }
+            }
+            if (firstNonNullElement == null) {
+                firstNonNullElement = element;
+            }
+        }
+        return firstNonNullElement;
+    }
+
+    @Override
+    public Iterable<Row> evaluate(TransactionContext txnCtx, Input<T>... args) {
+        assert args.length == 2 || args.length == 3 :
+            "Signature must ensure that there are either two or three arguments";
+
+        List<?> array = (List<?>) args[0].value();
+        if (array == null || array.isEmpty()) {
+            return () -> Collections.emptyIterator();
+        }
+        int dim = (Integer) args[1].value();
+        if (dim <= 0) {
+            throw new IllegalArgumentException("dimension must be greater than zero");
+        }
+        List<?> nestedArray = extractNestedArray(array, dim);
+        if (nestedArray == null || nestedArray.isEmpty()) {
+            return () -> Collections.emptyIterator();
+        }
+        boolean reversed = args.length == 3 && (Boolean) args[2].value();
+        int lastSubscript = reversed ? 1 : nestedArray.size();
+        int step = reversed ? -1 : 1;
+        return () -> new Iterator<>() {
+
+            final Object [] columns = new Object[]{ null };
+            final RowN row = new RowN(columns);
+            int idx = reversed ? nestedArray.size() : 1;
+
+            @Override
+            public boolean hasNext() {
+                return reversed ? idx >= lastSubscript : idx <= lastSubscript;
+            }
+
+            @Override
+            public Row next() {
+                if (!hasNext()) {
+                    throw new NoSuchElementException("no more rows");
+                }
+                columns[0] = idx;
+                idx += step;
+                return row;
+            }
+        };
+    }
```

Comment:
What do you think about extracting parts of the `evaluate` of `generate_series` and re-using it here?